dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,272 @@
+                if (backgroundTaskNameStr.find(backgroundTaskName) != std::wstring::npos)
+                {
+                    if (backgroundTaskNameStr.find(taskClsidStr) != std::wstring::npos)
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        throw winrt::hresult_invalid_argument(L"RegisterActivator has different clsid registered.");
+                    }
+                };
+                return false;
+            });
+
+            if (!taskRegistered)
+            {
+                builder = BackgroundTaskBuilder();
+                builder.Name(backgroundTaskName + taskClsidStr.c_str());
+
+                PushNotificationTrigger trigger{};
+                builder.SetTrigger(trigger);
+
+                auto builder5 = builder.try_as<winrt::IBackgroundTaskBuilder5>();
+                if (IsPackagedProcess() && builder5)
+                {
+                    builder5.SetTaskEntryPointClsid(taskClsid);
+                    winrt::com_array<winrt::IBackgroundCondition> conditions = details.GetConditions();
+                    for (auto condition : conditions)
+                    {
+                        builder.AddCondition(condition);
+                    }
+                }
+                else
+                {
+                    throw winrt::hresult_not_implemented();
+                }
+            }
+        }
+
+        if (WI_IsFlagSet(details.Option(), PushNotificationRegistrationOption::ComActivator))
+        {
+            {
+                auto lock = g_lock.lock();
+                // Define handle that will be set during background task execution
+                g_waitHandleForArgs = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
+                THROW_HR_IF_NULL(E_UNEXPECTED, g_waitHandleForArgs);
+            }
+
+            THROW_IF_FAILED(::CoRegisterClassObject(
+                taskClsid,
+                winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                CLSCTX_LOCAL_SERVER,
+                REGCLS_MULTIPLEUSE,
+                &cookie));
+        }
+
+        if (builder)
+        {
+            registeredTask = builder.Register();
+        }
+
+        return PushNotificationRegistrationToken{ cookie, registeredTask };
+    }
+
+    void PushNotificationManager::UnregisterActivator(PushNotificationRegistrationToken const& token, PushNotificationRegistrationOption const& option)
+    {
+        THROW_HR_IF_NULL(E_INVALIDARG, token);
+        if (WI_IsFlagSet(option, PushNotificationRegistrationOption::PushTrigger))
+        {
+            THROW_HR_IF_NULL(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), token.TaskRegistration());
+            for (auto task : BackgroundTaskRegistration::AllTasks())
+            {
+                if (task.Value().Name() == token.TaskRegistration().Name())
```

Comment:
Can cache the `token.TaskRegistration().Name()` to avoid redundant virtual calls.